cmd/poller/asup.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"encoding/json"
+	"goharvest2/cmd/harvest/version"
+	"goharvest2/pkg/conf"
+	"goharvest2/pkg/matrix"
+	"goharvest2/pkg/tree/node"
+	"io/ioutil"
+	"os/exec"
+	"strconv"
+	"strings"
+)
+
+type asupMessage struct {
+	Target   *targetInfo
+	Nodes    *instanceInfo
+	Volumes  *instanceInfo
+	Platform *platformInfo
+	Harvest  *harvestInfo
+}
+
+type targetInfo struct {
+	Version string
+	Model   string
+	Serial  string
+	Ping    float64
+}
+
+type instanceInfo struct {
+	Count      int64
+	DataPoints int64
+	PollTime   int64
+	ApiTime    int64
+	ParseTime  int64
+	PluginTime int64
+}
+
+type platformInfo struct {
+	OS       string
+	Arch     string
+	MemoryKb uint64
+	CPUs     uint8
+}
+
+type harvestInfo struct {
+	UUID        string
+	Version     string
+	Release     string
+	Commit      string
+	BuildDate   string
+	NumClusters uint8
+}
+
+func (p *Poller) doAsupMessage() (*matrix.Matrix, error) {
+
+	var (
+		msg *asupMessage
+		err error
+	)
+
+	if msg, err = p.buildAsupMessage(); err != nil {
+		logger.Error().Err(err).Msg("failed to build ASUP message")
+		return nil, err
+	}
+
+	if err = p.sendAsupMessage(msg); err != nil {
+		logger.Error().Err(err).Msg("failed to send ASUP message")
+		return nil, err
+	}
+
+	return nil, nil
+
+}
+
+// TODO: This function will be used to invoke harvest-asup(private repo)
+func (p *Poller) sendAsupMessage(msg *asupMessage) error {
+	logger.Info().Msgf("ASUP generatation started")
+
+	file, _ := json.MarshalIndent(msg, "", " ")
```

Comment:
more efficient to write directly to the file instead of holding the json in a string and writing the string. On the other hand, these json messages are small

something like this:
```
file, err := os.OpenFile("payload.json", os.O_CREATE, os.ModePerm)
defer file.Close()
encoder := json.NewEncoder(file) 
encoder.Encode(msg)
```

Reply:
done. added encoder with indentation and O_WRONLY.